File: src/core/projector/json-patch-stream.ts
/**
 * JSON-Patch streaming per RFC 6902 (vibe-kanban #3 / G5.F.3).
 *
 * The dispatcher's SSE stream previously re-sent the entire projected
 * state on every tick. For long-lived sessions this becomes expensive:
 * O(taskCount) JSON serialization per event even when only one card
 * changed.
 *
 * This module computes a minimal patch between two snapshots and
 * provides an `applyPatch` that mirrors the spec exactly enough for
 * our needs (we don't need `move`, `copy`, or `test` in v1 — only
 * `add`, `replace`, `remove`).
 *
 * Path encoding follows RFC 6901: `~` → `~0`, `/` → `~1`.
 *
 * v1 limitation: arrays are diffed as opaque values (no element-level
 * granularity). The TaskBoard projection uses a Record/Map shape at the
 * top level, so this is fine for the current consumers. If a future
 * consumer wants array-level diffing we can swap to `fast-json-patch`.
 */

export interface JsonPatchOpAdd {
	op: "add";
	path: string;
	value: unknown;
}
export interface JsonPatchOpReplace {
	op: "replace";
	path: string;
	value: unknown;
}
export interface JsonPatchOpRemove {
	op: "remove";
	path: string;
}
export type JsonPatchOp = JsonPatchOpAdd | JsonPatchOpReplace | JsonPatchOpRemove;

/** Per RFC 6901 §4 — escape `~` first, then `/`. */
function escapeToken(t: string): string {
	return t.replace(/~/g, "~0").replace(/\//g, "~1");
}
function unescapeToken(t: string): string {
	return t.replace(/~1/g, "/").replace(/~0/g, "~");
}

function isPlainObject(v: unknown): v is Record<string, unknown> {
	return (
		typeof v === "object" &&
		v !== null &&
		!Array.isArray(v) &&
		Object.getPrototypeOf(v) === Object.prototype
	);
}

function shallowEqual(a: unknown, b: unknown): boolean {
	if (a === b) return true;
	if (Array.isArray(a) && Array.isArray(b)) {
		if (a.length !== b.length) return false;
		return JSON.stringify(a) === JSON.stringify(b);
	}
	if (isPlainObject(a) && isPlainObject(b)) {
		return JSON.stringify(a) === JSON.stringify(b);
	}
	return false;
}

/**
 * Compute a JSON-Patch that turns `prev` into `next`. Operates
 * recursively on plain objects; arrays and scalars are diffed as whole
 * values (single `replace`).
 */
export function diffPatch(prev: unknown, next: unknown, basePath = ""): JsonPatchOp[] {
	if (shallowEqual(prev, next)) return [];

	// Type mismatch or non-object on one side → replace whole subtree.
	if (!isPlainObject(prev) || !isPlainObject(next)) {
		return [{ op: "replace", path: basePath || "", value: next }];
	}

	const ops: JsonPatchOp[] = [];
	const prevKeys = new Set(Object.keys(prev));
	const nextKeys = new Set(Object.keys(next));

	for (const key of nextKeys) {
		const path = `${basePath}/${escapeToken(key)}`;
		if (!prevKeys.has(key)) {
			ops.push({ op: "add", path, value: next[key] });
			continue;
		}
		const pv = prev[key];
		const nv = next[key];
		if (shallowEqual(pv, nv)) continue;
		if (isPlainObject(pv) && isPlainObject(nv)) {
			ops.push(...diffPatch(pv, nv, path));
		} else {
			ops.push({ op: "replace", path, value: nv });
		}
	}
	for (const key of prevKeys) {
		if (nextKeys.has(key)) continue;
		ops.push({ op: "remove", path: `${basePath}/${escapeToken(key)}` });
	}
	return ops;
}

function splitPath(path: string): string[] {
	if (path === "" || path === "/") return [];
	const parts = path.split("/").slice(1);
	return parts.map(unescapeToken);
}

/**
 * Apply a JSON-Patch to a deep clone of `doc` and return the new
 * document. The input is never mutated.
 */
export function applyPatch(doc: unknown, patch: JsonPatchOp[]): unknown {
	// Deep clone via JSON round-trip so callers can't observe partial
	// states if applyPatch throws halfway. This is the same trade-off
	// `fast-json-patch` makes in non-mutating mode.
	const next = JSON.parse(JSON.stringify(doc));
	for (const op of patch) {
		applyOne(next, op);
	}
	return next;
}

function applyOne(root: unknown, op: JsonPatchOp): void {
	const parts = splitPath(op.path);
	if (parts.length === 0) {
		// Patching the root is undefined behavior in our subset — skip.
		// (RFC 6902 §4.1 allows it for `replace`, but we don't use it.)
		return;
	}
	// Walk to the parent
	let cursor: unknown = root;
	for (let i = 0; i < parts.length - 1; i++) {
		if (!isPlainObject(cursor)) return;
		cursor = (cursor as Record<string, unknown>)[parts[i]];
	}
	if (!isPlainObject(cursor)) return;
	const last = parts[parts.length - 1];
	const parent = cursor as Record<string, unknown>;
	switch (op.op) {
		case "add":
		case "replace":
			parent[last] = op.value;
			break;
		case "remove":
			delete parent[last];
			break;
	}
}

File: crates/apohara-projector/src/json_patch_stream.rs
//! JSON-Patch streaming per RFC 6902 (vibe-kanban #3 / G5.F.3).
//!
//! The dispatcher's SSE stream previously re-sent the entire projected
//! state on every tick. For long-lived sessions this becomes expensive:
//! O(taskCount) JSON serialization per event even when only one card
//! changed.
//!
//! This module computes a minimal patch between two snapshots and
//! provides an `apply_patch` that mirrors the spec exactly enough for
//! our needs: only `add`, `replace`, `remove` (no `move`, `copy`, `test`).
//!
//! Path encoding follows RFC 6901: `~` -> `~0`, `/` -> `~1`.
//!
//! v1 limitation: arrays are diffed as opaque values (no element-level
//! granularity). The TaskBoard projection uses an object shape at the
//! top level, so this is fine for the current consumers.
//!
//! Ported from `src/core/projector/json-patch-stream.ts`.

use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

/// A single JSON-Patch operation. Tagged on `op` so it round-trips with
/// the TS shape verbatim.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "op", rename_all = "lowercase")]
pub enum JsonPatchOp {
    Add { path: String, value: Value },
    Replace { path: String, value: Value },
    Remove { path: String },
}

/// Per RFC 6901 section 4 — escape `~` first, then `/`.
fn escape_token(t: &str) -> String {
    t.replace('~', "~0").replace('/', "~1")
}

fn unescape_token(t: &str) -> String {
    // Order matters: `~1` first then `~0`, mirroring the TS code which
    // reverses the escape sequence.
    t.replace("~1", "/").replace("~0", "~")
}

/// Compute a JSON-Patch that turns `prev` into `next`. Operates
/// recursively on JSON objects; arrays and scalars are diffed as whole
/// values (single `replace`).
pub fn diff_patch(prev: &Value, next: &Value) -> Vec<JsonPatchOp> {
    let mut ops = Vec::new();
    diff_patch_inner(prev, next, "", &mut ops);
    ops
}

fn diff_patch_inner(prev: &Value, next: &Value, base_path: &str, ops: &mut Vec<JsonPatchOp>) {
    if prev == next {
        return;
    }

    let (prev_obj, next_obj) = match (prev.as_object(), next.as_object()) {
        (Some(p), Some(n)) => (p, n),
        _ => {
            // Type mismatch or non-object on one side -> replace whole subtree.
            ops.push(JsonPatchOp::Replace {
                path: base_path.to_string(),
                value: next.clone(),
            });
            return;
        }
    };

    // Iterate next-keys first (preserves insertion order from serde_json
    // when preserve_order isn't enabled; either way semantics match TS
    // because we always emit add/replace before remove).
    for (key, nv) in next_obj.iter() {
        let path = format!("{}/{}", base_path, escape_token(key));
        match prev_obj.get(key) {
            None => ops.push(JsonPatchOp::Add {
                path,
                value: nv.clone(),
            }),
            Some(pv) if pv == nv => {}
            Some(pv) => match (pv.as_object(), nv.as_object()) {
                (Some(_), Some(_)) => diff_patch_inner(pv, nv, &path, ops),
                _ => ops.push(JsonPatchOp::Replace {
                    path,
                    value: nv.clone(),
                }),
            },
        }
    }

    for key in prev_obj.keys() {
        if next_obj.contains_key(key) {
            continue;
        }
        ops.push(JsonPatchOp::Remove {
            path: format!("{}/{}", base_path, escape_token(key)),
        });
    }
}

fn split_path(path: &str) -> Vec<String> {
    if path.is_empty() || path == "/" {
        return Vec::new();
    }
    // Skip the leading empty segment from the leading `/`.
    path.split('/').skip(1).map(unescape_token).collect()
}

/// Apply a JSON-Patch to a clone of `doc` and return the new document.
/// The input is never mutated.
pub fn apply_patch(doc: &Value, patch: &[JsonPatchOp]) -> Value {
    let mut next = doc.clone();
    for op in patch {
        apply_one(&mut next, op);
    }
    next
}

fn apply_one(root: &mut Value, op: &JsonPatchOp) {
    let path = match op {
        JsonPatchOp::Add { path, .. }
        | JsonPatchOp::Replace { path, .. }
        | JsonPatchOp::Remove { path } => path,
    };
    let parts = split_path(path);
    if parts.is_empty() {
        // Patching the root is undefined in our subset — skip (matches TS).
        return;
    }

    // Walk to the parent object.
    let (last, ancestors) = parts.split_last().expect("parts non-empty checked above");
    let mut cursor: &mut Value = root;
    for seg in ancestors {
        match cursor.as_object_mut() {
            Some(obj) => match obj.get_mut(seg) {
                Some(child) => cursor = child,
                None => return,
            },
            None => return,
        }
    }
    let parent: &mut Map<String, Value> = match cursor.as_object_mut() {
        Some(o) => o,
        None => return,
    };
    match op {
        JsonPatchOp::Add { value, .. } | JsonPatchOp::Replace { value, .. } => {
            parent.insert(last.clone(), value.clone());
        }
        JsonPatchOp::Remove { .. } => {
            parent.remove(last);
        }
    }
}
